File: solutions/python/luhn/3/luhn.py
class Luhn:
    def __init__(self, card_num):
        self.card_num = card_num

    def valid(self):
        stripped = self.card_num.replace(" ", "")
        if len(stripped) <= 1 or not stripped.isdigit():
            return False
        digits = [int(d) for d in stripped]
        for i in range(len(digits) - 2, -1, -2):
            digits[i] *= 2
            if digits[i] > 9:
                digits[i] -= 9
        return sum(digits) % 10 == 0

File: solutions/ruby/luhn/2/luhn.rb
class Luhn
  def self.valid?(number)
    stripped = number.gsub(" ", "")
    return false if stripped.length <= 1
    return false if stripped.match?(/\D/)
    digits = stripped.chars.map(&:to_i)
    (digits.length - 2).step(0, -2) do |i|
      digits[i] *= 2
      digits[i] -= 9 if digits[i] > 9
    end
    digits.sum % 10 == 0
  end
end